### PR TITLE
WFIRST Cycle 8 Updates 

### LINKED ISSUE #296: WFI Detector Field Coverage
Blocking: #294
@shanosborne  and I found that the current reference data does not cover all pixels in a given WFI detector. Each detecter has 5 field points. Four at the corners and one at the center:
WFI_Points

We interpolate between these points to determine the field dependance of the PSF at a given point. The reference data does not cover all corners which leads to an inability to interpolate aberration terms for border pixels. The reference data is indented by a 4 pixel margin:
WFI_Points_corner

When the user selects a pixel outside the area covered by the interpolation, an exception is raised:
`RuntimeError("Attempted to get aberrations for an out-of-bounds field point")`

The reasoning behind the placement of the field points can be found in this code comment:
https://github.com/spacetelescope/webbpsf/blob/master/webbpsf/wfirst.py#L197%03-L202

The question is now how to handle this error? Should we inform the user that they picked a valid pixel and that it is out of bound of the ref data? Should we extrapolate or estimate  (closest point approximation)?

### PR DESCRIPTION
**Pull request by robelgeda**
closes #296, #285

This is the final draft of WFIRST Cycle 8 updates. The  Following changes have been made:
NASA Ref Info: https://wfirst.gsfc.nasa.gov/science/WFIRST_Reference_Information.html

# Updates

### webbpsf:

-  **New** `WFIPupilController` class: The current cycle data splits up the pupil images by SCA. For this reason we need to make a controller to swap in the correct pupil each time the detector is changed. Please note that there are no changes to the user API, all functions remain available and interface with the new class when called.

- **Update** the detector aberrations file path to point to `wim_zernikes_cycle8.csv`

- **New** tests for the `WFIPupilController` class under `test_WFI_pupil_controller`

- **Update** filter names in `test_wfirst.py`

- **New**: the pupil mask is auto selected based on the filter selection as opposed to the wavelength range. 

- **New**: `test_WFI_fwhm` test function [@Skyhawk172]

- **Closes** #296. The `FieldDependentAberration` class uses `np.clip(self.field_position, 4, 4092)` to estimate aberration terms for margin field points (which are out of the data's interpolation zone). The `field_point` property does not allow the user to select field points outside of the detector so it is safe to rely on this one line of code. 

- **Update** out of bound field point tests `test_wfirst.py`

### webbpsf-data-source:
- **Re-named** `webbpsf-data/WFI/sources/cycle7/wim_zernikes_cycle7.csv` -> `webbpsf-data/WFI/sources/cycle7/ori_wim_zernikes_cycle7.csv` to indicate its the original version of the file.

- **Archived**  cycle 7 zernikes by moving `webbpsf-data/WFI/wim_zernikes_cycle7.csv` -> `webbpsf-data/WFI/sources/cycle7/wim_zernikes_cycle7.csv`

- **Archived** `webbpsf-data/WFIRST_SRR_WFC_Pupil_Mask_Longwave_2048.fits` ->
`webbpsf-data/WFI/sources/cycle7/WFIRST_SRR_WFC_Pupil_Mask_Longwave_2048.fits`

- **Archived**  `webbpsf-data/WFIRST_SRR_WFC_Pupil_Mask_Shortwave_2048.fits` ->
`webbpsf-data/WFI/sources/cycle7/WFIRST_SRR_WFC_Pupil_Mask_Shortwave_2048.fits`

- **Archived** `webbpsf-data/WFI/filters` -> `webbpsf-data/WFI/filters/sources/cycle7/filters`

- **Archived** `webbpsf-data/WFI/filters.tsv` -> `webbpsf-data/WFI/filters/sources/cycle7/filters.tsv`

- **Archived** `webbpsf-data/WFI/upscaled_HST_OPD.fits` -> `webbpsf-data/WFI/sources/cycle8/original_upscaled_HST_OPD.fits`

- **New** folder with pupils:  `webbpsf-data/WFI/pupils`. Masked and unmasked pupil images provided for each SCA.
    - Resized pupil images to a standard pixel scale.

- **Updated** Zernike Coefficients

- **Updated** Filter names (see section below for details)

- **Updated** Detailed filter throughputs from **Pandeia 1.4.2** 

- **Updated** Wavelength range now extended to cover the 0.48 - 2.0 µm range.

- **Updated** `upscaled_HST_OPD.fits` to match pupil image pixel scale.

# Checks

- [x] Filters bandpasses: (note we are using filter bandpasses from Pandeia)
    - [x] F087
    - [x] F106
    - [x] F129
    - [x] F158
    - [x] F184
    - [x] F146
    - [x] F062

- [x] Instrument layout: Instrument layout has changed and the data matches the new layout 
 
- [x] Zernike coefficients 

- [x] New pupil images are compatible with the existing codebase. Investigate how the new individualized pupil images differ from the singular image in the past. (Dimensions of the image etc...)

# Notable Changes to the WFIRST Design  

- **Wavelength range was updated to cover 0.48 - 2.0 µm range.** 
- Masked and unmasked pupil images provided for each SCA. 
- Placement of the gap in the WFI instrument was moved.
- Filters have been renamed. The leading letter has been replaced with "F":
Screen Shot 2019-07-23 at 10 54 25 AM

- The instrument layout plot was updated to reflect the correct positioning of SCA numbers. (The cycle 7 plot was not correct)

WFIRST_Simulated_WFI

- Explanation of how to transform between coordinate systems:

> On slide 3 [of the powerpoint], the FPA sky projection is in the mm (length space) made by zemax, not in angular space. If you want transfer the mm space into angular space, the following is the angular sign convention used in ZEMAX: angles are positive when they have positive slope relative to the local optical axis. (here the local optical axis is +z, which point towards Primary Mirror). So, FPA 7 on the SKY Project has +x coordinates, but the slop of it's beam (direction from sky projection surface to Pupil or Primary Mirror) relative to local optical axis (+z towards primary) is negative, so FPA 7 has negative Gield angle in angular space.  And slides page 4 has  the plot of Field of view in angular space (from zemax) and it's mapping to the FPA local surface (not sky projection). so in FPA local surface, FPA 7 is on the negative side of x-axis (left hand coordinate), mapping to sky projection surface, FPA 7 is on the positive side of x-axis (right hand coordinate, change from left hand to right hand is because there are odd times of reflection), then mapping to field of view angular space, FPA 7 is at the negative side of x-axis. So top row should be "16,13,10,1,4,7".

>  Slides 4 is the mapping relation from field of view angular space to FPA local coordinate space (produce by zemax), you can see a point (A) with negative x angle has a negative x in FPA local coordinate. In Slide 3,  point A (has negative x-angle and negative FPA local x coordinate) is at the positive size of x-axis on the sky projection surface.

- Filters have detailed throughput data:
Screen Shot 2019-07-23 at 6 54 01 PM

### CODE REVIEW

**Comment by robelgeda** [comment]
I refactored the pupil code and created a `WFIPupilController` class. The purpose of the class is to extract and concentrate the pupil and mask selection code into a single class. This was motivated by increased complexity in pupil file selection. The pupil now depends on the selected detector and the `pupil_mask` flag. 
The last commit was to test if the refactor works for the cycle 7 data. Travis passes the commit indicating the refactor is successful. The next step will be to gear the code towards cycle 8 data.

File: webbpsf/wfirst.py [hunk 1]
```diff
@@ -362,15 +405,102 @@ def _validate_config(self, **kwargs):
             # If the user has set the pupil to a custom value, let them worry about the
             # correct shape it should have
             pass
+
+    def user_pupil_mask_override(self, set_pupil_mask_on):
+        if set_pupil_mask_on is not None:
```
Comment (mperrin) [inline webbpsf/wfirst.py]: The `if set_pupil_mask_on is not None:` seems unnecessary. If you take that line out, then if called with a None, it would just go to the TypeError exception stating that the parameter must be boolean. That seems like good behavior. 

As written now, if passed a None this function would silently do nothing and raise no error. Is that the desired behavior?

File: webbpsf/wfirst.py [hunk 2]
```diff
@@ -276,73 +276,116 @@ def _get_fits_header(self, result, options):
...
-        Parameters
-        -----------
-        set_pupil_mask_on : bool or None
-            Set to True or False to force using or not using the cold pupil mask,
-            or to None for the automatic behavior.
-        """
-        pixelscale = 110e-3  # arcsec/px, WFIRST-AFTA SDT report final version (p. 91)
-        super(WFI, self).__init__("WFI", pixelscale=pixelscale)
+    def __init__(self, wfi):
+        self.wfi = wfi
```
[the remark this answers is not in the document]
Comment (robelgeda) [reply webbpsf/wfirst.py]: I agree with this. I made some changes.

File: webbpsf/wfirst.py [hunk 3]
```diff
@@ -276,73 +276,116 @@ def _get_fits_header(self, result, options):
...
-        self.pupil = self._unmasked_pupil_path
-        if set_pupil_mask_on is not None:
-            if isinstance(set_pupil_mask_on, bool):
+    @property
+    def pupil(self):
+        return self._pupil
+
+    @pupil.setter
+    def pupil(self, value):
+        self._pupil = value
```
[the remark this answers is not in the document]
Comment (robelgeda) [reply webbpsf/wfirst.py]: We need the parameter because setting the class variable to a value will change the value of the class variable instead of the value of the object that its pointing to. For example: 
```
In [2]: class A:
   ...:     _x = None
   ...:     @property
   ...:     def x(self):
   ...:         return self._x
   ...:     @x.setter
   ...:     def x(self, v):
   ...:         self._x = v
   ...:

In [3]: class B:
   ...:     def __init__(self):
   ...:         self.a = A()
   ...:         self.y = self.a.x
   ...:

In [4]: B()
Out[4]: <__main__.B at 0x10d96e7b8>

In [5]: b = B()

In [6]: b.y

In [7]: b.y = 22

In [8]: b.a.x

In [9]: b.a.x = 44

In [10]: b.y
Out[10]: 22

In [11]:
```

File: webbpsf/wfirst.py [hunk 4]
```diff
@@ -276,73 +276,116 @@ def _get_fits_header(self, result, options):
...
-        ]
-        self.pupilopd = self.opd_list[-1]
+        self._update_pupil()
 
-    def _validate_config(self, **kwargs):
+    def update_pupil_path(self, detector):
+        """
+        Update the masked and unmasked pupil paths according to the SCA selected
+        """
+        detector = detector or self.wfi.detector
```
[the remark this answers is not in the document]
Comment (robelgeda) [reply webbpsf/wfirst.py]: That is correct. Since we are removing the reference to the wfi, I removed that line.

File: webbpsf/wfirst.py (hunk 1 above)
Comment (robelgeda) [reply webbpsf/wfirst.py]: I didn't think this one through

**Comment by robelgeda** [comment]
@mperrin @shanosborne : This PR is ready for review and/or merger.

**Comment by mperrin** [comment]
Hi @robelgeda, ok I will take a look and review. Can you please point me at where I can find the updated WebbPSF data files including the modifications for cycle 8? Thanks. 

We should put a copy of those in the Box folder I created yesterday, and update the Travis tests to retrieve from there, in order to get the tests passing on here.

**Comment by Skyhawk172** [comment]
Okay - I am just getting around to updating my local repos and remembering everything about Git. I will take a look too.

**Comment by mperrin** [comment]
Ok, @Skyhawk172 please check in with @robelgeda as he’s been continuing work on this, as partly discussed on Slack rather than here.

**Comment by robelgeda** [comment]
Sure thing, I am currently working on a sprint but I can make time next week to tag up with you @Skyhawk172 if that works?

**Comment by Skyhawk172** [comment]
Hi Robel - I'll be busy all of next week with the advanced Python training. I'll try to catch up on your updates as I can and hopefully we can touch base the following week.

File: webbpsf/wfirst.py [hunk 5]
```diff
@@ -342,35 +502,56 @@ def _validate_config(self, **kwargs):
...
+        self._pupil_controller.validate_pupil(self.filter, **kwargs)
         super(WFI, self)._validate_config(**kwargs)
 
+    @property
+    def detector(self):
+        """Detector selected for simulated PSF
+        Used in calculation of field-dependent aberrations. Must be
+        selected from detectors in the `detector_list` attribute.
+        """
+        return self._detector
```
Comment (mperrin) [inline webbpsf/wfirst.py]: I believe this `detector` function is identical to that in the superclass, so the definition doesn't need to be repeated here.  In a subclass, you can override only a property setter without having to redefine the getter function, by accessing the property in the parent class with the class name. 

In other words, just below here you can change 
```
    @detector.setter
```
into
```
    @SpaceTelescopeInstrument.detector.setter
```
and not have to define `@detector` redundantly here.

File: webbpsf/wfirst.py [hunk 6]
```diff
@@ -342,35 +502,56 @@ def _validate_config(self, **kwargs):
...
+
     @property
     def pupil_mask(self):
-        return self._pupil_mask
+        return self._pupil_controller.pupil_mask
+
+    @property
+    def filter(self):
+        """Currently selected filter name (e.g. F200W)"""
+        return self._filter
```
Comment (mperrin) [inline webbpsf/wfirst.py]: Similar as above comment. No need to redefine the filter getter function here with no change; you can override only the setter function by invoking the superclass name.

File: webbpsf/wfirst.py [hunk 7]
```diff
@@ -384,29 +565,19 @@ def pupil_mask(self, name):
...
-        else:
-            raise ValueError("Pupil mask setting is not valid or empty.")
-        self._pupil_mask = name
+        self._pupil_controller.pupil_mask = name
 
     def _addAdditionalOptics(self, optsys, **kwargs):
         return optsys, False, None
 
+    @property
+    def _unmasked_pupil_path(self):
```
Comment (mperrin) [inline webbpsf/wfirst.py]: What are these hidden properties needed for? Is there a way the code could be simplified to not need these defined on the WFI class in duplication of the pupil controller class?

File: webbpsf/wfirst.py [hunk 8]
```diff
@@ -285,10 +463,6 @@ class WFI(WFIRSTInstrument):
              simulations, and uses several approximations (e.g. for
              mirror polishing errors, which are taken from HST).
     """
-    # "The H158, F184 and W149 filters and the grism are mounted with proximate cold pupil masks"
-    # from the final draft of the SDT report, page 92, table 3-2
-    UNMASKED_PUPIL_WAVELENGTH_MIN, UNMASKED_PUPIL_WAVELENGTH_MAX = 0.760e-6, 1.454e-6
-    MASKED_PUPIL_WAVELENGTH_MIN, MASKED_PUPIL_WAVELENGTH_MAX = 1.380e-6, 2.000e-6
 
     def __init__(self, set_pupil_mask_on=None):
```
[the remark this answers is not in the document]
Comment (ariedel) [reply webbpsf/wfirst.py]: I still need it (or something like it), because I need to generate monochromatic PSFs in the overlapping wavelength range (1.38 microns - 1.454 microns) with AND without the pupil mask, so I need a way to explicitly control that.

[the remark this answers is not in the document]
Comment (ariedel) [reply webbpsf/wfirst.py]: On further inspection and feedback from @rgeda, our code is setting the pupil_mask attribute of the class object, not setting this flag in the init script. So I do get this functionality, in another way.

tl;dr: I actually don't have any need for this code.

**Comment by robelgeda** [comment]
@mperrin @shanosborne : The tests on the new dataset works fine. The last thing left to do is to do an actual release of the dataset and then point Travis to it (currently I am using a release candidate). Please let me know once the new dataset goes officially live.

**Comment by Skyhawk172** [comment]
In my attempt to investigate the WFI PSFs using the different pupils, I found a bug when setting pupilopd = None.  When doing so, the code reverts to the amplitude/pupil files to determine the pixelscale and looks for specific keywords ('PIXELSCL', 'PUPLSCAL'), but they are not present in the pupil files.  We could either create a new PUPLSCAL keyword in the pupil files from the existing CDELT/PUPILMAG keywords, or change the code in " _find_pixelscale_in_headers()" to be able to handle CDELT/PUPILMAG.

**Comment by mperrin** [comment]
Good catch! I recommend adding new keywords to the FITS headers with appropriate value. That’s the cleanest fix. 

PUPILMAG is not a standard keyword and basically doesn’t exist in any other FITS files but these, so it  it doesn’t seem efficient to add that as another option in the code. It’s more elegant to ensure the data files for WFIRST have consistent keywords with those for JWST, especially since we’re modifying those files already to do the rescaling to a common pixel scale.

**Comment by robelgeda** [comment]
I am so busy this week, I can only get to this next week.

**Comment by robelgeda** [comment]
@mperrin @Skyhawk172 I updated the files. Please let me know if they look good.

**Comment by Skyhawk172** [comment]
For (historical) completeness, let me just add here that we also found out that the units between PIXELSCAL and PUPLSCAL were different but that poppy was treating them both as meter/pixel.  So we decided to overwrite the PIXELSCAL values with the new PUPLSCAL values (in m/pix) and remove PUPLSCAL from the header altogether.  Also added comments in header to help reconstruct what happened and provided a new unit test (test_WFI_fwhm). No code changes in poppy.
